ISPN-6488 Convert CacheRpcCommands to use a pre-computed ByteStrings

tristantarrant opened the pull request:
https://issues.jboss.org/browse/ISPN-6488

WARNING: this incarnation is backwards incompatible since it changes the CacheRpcCommand to _only_ take a ByteString instead of a String. I did this to make sure I didn't miss any existing internal commands, but I will re-add the String constructors once we are happy

tristantarrant commented:
We might NOT want the String constructors.

galderz commented on core/src/main/java/org/infinispan/util/ByteString.java:
```diff
@@ -0,0 +1,90 @@
+ * @author Tristan Tarrant
+ * @since 9.0
+ */
+public class ByteString {
```
Why do you need this wrapper class? Can't you just work with byte[] directly? What's stopping you?

tristantarrant replied:
To abstract the logic. As @slaskawi says below, we might decide we want to use a different encoding, hashing, etc down the line

galderz replied:
Different encoding/hashing...etc can be achieved with a static class that gets the byte[] and produces what you want. I'm not 100% sure we really need a wrapper here, but feel free to go ahead with it, we can always change later.

The only real problem we have with byte[] is equality and hashcode but that's already sorted out with the equivalence functions that are separated out.

slaskawi commented on core/src/main/java/org/infinispan/util/ByteString.java:
```diff
@@ -0,0 +1,90 @@
+
+      @Override
+      public ByteString readObject(ObjectInput input) throws IOException, ClassNotFoundException {
+         short l = input.readByte();
```
Could we use `len` here or something more descriptive? `l` looks like '1' and it doesn't look good in `new byte[1]` :)

slaskawi commented on core/src/main/java/org/infinispan/util/ByteString.java:
```diff
@@ -0,0 +1,90 @@
+import java.util.Collections;
+import java.util.Set;
+
+/**
```
Could you please add some meaningful javadoc here?

tristantarrant replied:
Done

slaskawi commented:
If I understand this correctly - we are using `ByteString`s everywhere to make everybody aware that some chars might be lost during the conversion? That's the argument why we shouldn't use String all around and convert them to `ByteString`s just before sending through JGroups?

tristantarrant commented:
@slaskawi, no the argument is that String.writeUTF is an extremely expensive op that currently is performed for every CacheRpcCommand. Precomputing a byte[] representation of it once (at Cache construction) ensures we skip that.

tristantarrant commented:
@slaskawi  addressed your comments
@galderz I still prefer encapsulation. But if anybody else has arguments against the ByteString class please voice them

slaskawi commented:
Looks good to me.

slaskawi commented:
I see CI has failed; let's rerun it.

pruivo commented on core/src/main/java/org/infinispan/util/ByteString.java:
```diff
@@ -0,0 +1,93 @@
+            return EMPTY;
+
+         byte[] b = new byte[len];
+         input.read(b);
```
```readFully``` is more correct than ```read```

tristantarrant replied:
Done

pruivo commented on core/src/main/java/org/infinispan/factories/GlobalComponentRegistry.java:
```diff
@@ -84,6 +85,7 @@
    final Collection<ModuleLifecycle> moduleLifecycles;
 
    final ConcurrentMap<String, ComponentRegistry> namedComponents = new ConcurrentHashMap<String, ComponentRegistry>(4);
+   final ConcurrentMap<ByteString, ComponentRegistry> byteNamedComponents = new ConcurrentHashMap<ByteString, ComponentRegistry>(4);
```
you can keep only the ```byteNamedComponents``` map. ```getNamedComponentRegistry(String)``` is only invoked when a cache is started or stopped, so you can afford the cost of converting the ```String``` => ```ByteString``` in that cases.

tristantarrant replied:
Done

pruivo commented on core/src/main/java/org/infinispan/util/ByteString.java:
```diff
@@ -0,0 +1,93 @@
+
+   @Override
+   public String toString() {
+      return new String(b, CHARSET);
```
I'm a little concern about this. wouldn't be better to cache the string too?

danberindei replied:
AFAICT this is only used in `LogFactory.pushNDC(ByteString, boolean)`, and that is only used for `StateRequestCommand`, and `StateResponseCommand`, so it shouldn't be a problem. 

And even those usages could be eliminated by calling `LogFactory.pushNDC()` from the inbound invocation handler for all commands, instead of `InvocationContextInterceptor` for some and `perform()` for others.

danberindei commented on core/src/main/java/org/infinispan/commands/tx/VersionedCommitCommand.java:
```diff
@@ -20,14 +21,14 @@
 
    public VersionedCommitCommand() {
-      super("");
+      super(ByteString.emptyString());
```
Might as well use `null`, almost all commands use `super(null)` in the "CommandIdUniquenessTest constructor".

tristantarrant replied:
Done

danberindei commented on query/src/main/java/org/infinispan/query/indexmanager/AbstractUpdateCommand.java:
```diff
@@ -70,14 +71,15 @@ public boolean isReturnValueExpected() {
    public void fetchExecutionContext(CommandInitializer ci) {
-      if (ci.getCacheManager().cacheExists(cacheName)) {
-         Cache cache = ci.getCacheManager().getCache(cacheName);
+      String name = cacheName.toString();
```
I guess I spoke to soon about `ByteString.toString()` not being used... I guess here we need a way for the per-cache inbound invocation handler to pass the `Cache` to the command.

tristantarrant replied:
I am caching the String on first invocation of toString() now

danberindei commented:
@tristantarrant My ideal solution would be to remove the cache name from `CacheRpcCommand` altogether, and to have only the `RpcManager` and the `PerCacheInboundInvocationHandler` deal with cache references (be they actual names, hashes, or something else). But that would be a "slightly" bigger change :) 

At this point, I like having a `ByteString` reference instead of a raw `byte[]`, I'd just use static methods for serialization/deserialization instead of a full externalizer.

tristantarrant commented:
Dropped the Externalizer and introduced the two static methods.

pruivo commented:
looks good to me :+1:

danberindei commented on core/src/main/java/org/infinispan/marshall/core/ExternalizerTable.java:
```diff
@@ -106,6 +106,7 @@
 import org.infinispan.transaction.xa.recovery.RecoveryAwareDldGlobalTransaction;
 import org.infinispan.transaction.xa.recovery.RecoveryAwareGlobalTransaction;
 import org.infinispan.transaction.xa.recovery.SerializableXid;
+import org.infinispan.util.ByteString;
```
No longer needed ;)

tristantarrant commented:
@danberindei removed the import

danberindei commented:
Integrated, thanks Tristan!
